### PR TITLE
Interactivity: Strict type checking

### PR DESCRIPTION
**Pull request by sirreal**
## What?

Typecheck more of the interactivity package.

Part of #60219

## Why?

The package is partially TypeScript and has relaxed strictness that reduces the quality of the produced type declarations.

<details>
<summary>Types get better!</summary>

```patch
diff --git a/packages/interactivity/build-types/directives.d.ts b/../trunk/packages/interactivity/build-types/directives.d.ts
index 9b8fba02b7b..4cab5535aa1 100644
--- a/packages/interactivity/build-types/directives.d.ts
+++ b/../trunk/packages/interactivity/build-types/directives.d.ts
@@ -1,2 +1,2 @@
-declare function _default(): void;
+declare const _default: () => void;
 export default _default;
diff --git a/packages/interactivity/build-types/hooks.d.ts b/../trunk/packages/interactivity/build-types/hooks.d.ts
index bb5d8af5dc2..d2be46b216f 100644
--- a/packages/interactivity/build-types/hooks.d.ts
+++ b/../trunk/packages/interactivity/build-types/hooks.d.ts
@@ -1,10 +1,10 @@
 /**
  * External dependencies
  */
-import { h as createElement } from 'preact';
+import { h as createElement, type ComponentChildren } from 'preact';
 import type { VNode, Context, RefObject } from 'preact';
 interface DirectiveEntry {
-    value: string | Object;
+    value: string | object;
     namespace: string;
     suffix: string;
 }
@@ -17,11 +17,17 @@ interface DirectiveArgs {
     /**
      * Props present in the current element.
      */
-    props: Object;
+    props: {
+        children?: ComponentChildren;
+    };
     /**
      * Virtual node representing the element.
      */
-    element: VNode;
+    element: VNode<{
+        class?: string;
+        style?: string | Record<string, string | number>;
+        content?: ComponentChildren;
+    }>;
     /**
      * The inherited context.
      */
@@ -33,7 +39,7 @@ interface DirectiveArgs {
     evaluate: Evaluate;
 }
 interface DirectiveCallback {
-    (args: DirectiveArgs): VNode | void;
+    (args: DirectiveArgs): VNode | null | void;
 }
 interface DirectiveOptions {
     /**
@@ -46,7 +52,7 @@ interface DirectiveOptions {
 }
 interface Scope {
     evaluate: Evaluate;
-    context: Context<any>;
+    context: object;
     ref: RefObject<HTMLElement>;
     attributes: createElement.JSX.HTMLAttributes;
 }
@@ -76,7 +82,7 @@ export declare const getContext: <T extends object>(namespace?: string) => T;
  * @return Element representation.
  */
 export declare const getElement: () => Readonly<{
-    ref: HTMLElement;
+    ref: HTMLElement | null;
     attributes: createElement.JSX.HTMLAttributes<EventTarget>;
 }>;
 export declare const getScope: () => Scope;
diff --git a/packages/interactivity/build-types/vdom.d.ts b/../trunk/packages/interactivity/build-types/vdom.d.ts
index 24000973702..e1b76d3defa 100644
--- a/packages/interactivity/build-types/vdom.d.ts
+++ b/../trunk/packages/interactivity/build-types/vdom.d.ts
@@ -1,8 +1,12 @@
+/**
+ * External dependencies
+ */
+import { type VNode } from 'preact';
 export declare const hydratedIslands: WeakSet<object>;
 /**
  * Recursive function that transforms a DOM tree into vDOM.
  *
- * @param {Node} root The root element or node to start traversing on.
- * @return {import('preact').VNode[]} The resulting vDOM tree.
+ * @param root The root element or node to start traversing on.
+ * @return The resulting vDOM tree.
  */
-export declare function toVdom(root: any): any[];
+export declare function toVdom(root: Node): [string | VNode | null, Node | null];
```

</details>

## How?

Migrate more file to TypeScript and fix type errors.

## Testing Instructions

CI should be sufficient.

### CODE REVIEW

File: packages/interactivity/tsconfig.json [hunk 1]
```diff
@@ -4,8 +4,11 @@
 	"compilerOptions": {
 		"rootDir": "src",
 		"declarationDir": "build-types",
-		"checkJs": false,
-		"strict": false
+
+		"strictNullChecks": false,
+		"strictPropertyInitialization": false,
+
+		"noImplicitAny": false
```
Comment (sirreal) [inline packages/interactivity/tsconfig.json]: We'll gradually want to remove these so all strictness is enabled.

File: packages/interactivity/src/vdom.ts [hunk 2]
```diff
@@ -38,33 +40,49 @@ export const hydratedIslands = new WeakSet();
 /**
  * Recursive function that transforms a DOM tree into vDOM.
  *
- * @param {Node} root The root element or node to start traversing on.
- * @return {import('preact').VNode[]} The resulting vDOM tree.
+ * @param root The root element or node to start traversing on.
+ * @return The resulting vDOM tree.
  */
-export function toVdom( root ) {
+export function toVdom( root: Node ): [ string | VNode | null, Node | null ] {
```
Comment (sirreal) [inline packages/interactivity/src/vdom.ts]: I'm not happy with the changes here…

Comment (cbravobernal) [reply packages/interactivity/src/vdom.ts]: `toVdom` is an internal function. So TS here is not critical. Don't let this block your PR 😄

**Comment by sirreal** [comment]
This needs a changelog entry. It's strict null type checked now so the generated types should be better.

I think there was an issue around some possible nulls not being reflected in types (around getElement?) that should be fixed now.

**Comment by cbravobernal** [comment]
> This needs a changelog entry. It's strict null type checked now so the generated types should be better.

Done in https://github.com/WordPress/gutenberg/pull/59865/commits/982606e2cb8b93378817d5a77587f3881966124e

File: packages/interactivity/src/vdom.ts [hunk 3]
```diff
@@ -39,81 +39,100 @@ export const hydratedIslands = new WeakSet();
...
 				} else {
-					let [ ns, value ] = nsPathRegExp
+					const [ ns, value ] = nsPathRegExp
 						.exec( attributes[ i ].value )
 						?.slice( 1 ) ?? [ null, attributes[ i ].value ];
+					let parsedValue: Record< string, unknown > | null = null;
 					try {
-						value = JSON.parse( value );
+						parsedValue = value ? JSON.parse( value ) : null;
 					} catch ( e ) {}
```
Comment (sirreal) [inline packages/interactivity/src/vdom.ts]: This is a regression. I was only thinking about the namespace case.

Comment (sirreal) [reply packages/interactivity/src/vdom.ts]: Should be fixed now.

**Comment by sirreal** [comment]
This should fix the types that were a problem in https://github.com/WordPress/gutenberg/issues/59757

File: packages/interactivity/CHANGELOG.md [hunk 4]
```diff
@@ -2,10 +2,13 @@
 
 ## Unreleased
 
+### Enhancements
+
+-   Strict type checking. ([#59865] (https://github.com/WordPress/gutenberg/pull/59865/))
```
Comment (sirreal) [inline packages/interactivity/CHANGELOG.md]:
```suggestion
-   Strict type checking: fix some missing nulls in published types ([#59865](https://github.com/WordPress/gutenberg/pull/59865/)).
```

File: packages/interactivity/src/vdom.ts [hunk 5]
```diff
@@ -32,88 +32,109 @@ const directiveParser = new RegExp(
...
+		const children: Array< ComponentChild > = [];
+		const directives: Array<
+			[ name: string, namespace: string | null, value: unknown ]
+		> = [];
 		let ignore = false;
 		let island = false;
 
 		for ( let i = 0; i < attributes.length; i++ ) {
-			const n = attributes[ i ].name;
+			const attributeName = attributes[ i ].name;
```
Comment (cbravobernal) [inline packages/interactivity/src/vdom.ts:90]: 🚀  Thanks for making this easier to read.

File: packages/interactivity/src/vdom.ts [hunk 6]
```diff
@@ -32,88 +32,109 @@ const directiveParser = new RegExp(
...
+				if ( attributeName === ignoreAttr ) {
 					ignore = true;
 				} else {
-					let [ ns, value ] = nsPathRegExp
-						.exec( attributes[ i ].value )
-						?.slice( 1 ) ?? [ null, attributes[ i ].value ];
+					const regexResult = nsPathRegExp.exec(
+						attributes[ i ].value
+					)?.groups;
+					const namespace = regexResult?.nameSpace ?? null;
```
Comment (sirreal) [inline packages/interactivity/src/vdom.ts]: Ugh…
```suggestion
					const namespace = regexResult?.namespace ?? null;
```

**Review by cbravobernal — APPROVED** [review]
We need this PR (we can iterate later if needed) merged asap so it does not conflict with future developments.

Great work @sirreal

File: packages/interactivity/src/vdom.ts (hunk 6 above)
Comment (cbravobernal) [reply packages/interactivity/src/vdom.ts]: Not the first one, not the last one

File: packages/interactivity/tsconfig.json [hunk 7]
```diff
@@ -4,8 +4,8 @@
 	"compilerOptions": {
 		"rootDir": "src",
 		"declarationDir": "build-types",
-		"checkJs": false,
-		"strict": false
+
+		"noImplicitAny": false
```
Comment (sirreal) [inline packages/interactivity/tsconfig.json:8]: ~I think this can be removed now.~

File: packages/interactivity/src/vdom.ts [hunk 8]
```diff
@@ -32,88 +32,109 @@ const directiveParser = new RegExp(
...
 				} else {
-					let [ ns, value ] = nsPathRegExp
-						.exec( attributes[ i ].value )
-						?.slice( 1 ) ?? [ null, attributes[ i ].value ];
+					const regexResult = nsPathRegExp.exec(
+						attributes[ i ].value
+					)?.groups;
+					const namespace = regexResult?.namespace ?? null;
+					let value: any =
+						regexResult?.value ?? attributes[ i ].value;
```
Comment (sirreal) [inline packages/interactivity/src/vdom.ts]: A bit better:

```suggestion
					const regexCaptureGroups = nsPathRegExp.exec(
						attributes[ i ].value
					)?.groups;
					const namespace = regexCaptureGroups?.namespace ?? null;
					let value: any =
						regexCaptureGroups?.value ?? attributes[ i ].value;
```

File: packages/interactivity/tsconfig.json (hunk 7 above)
Comment (sirreal) [reply packages/interactivity/tsconfig.json:8]: Wrong, this isn't noImplicitAny yet.

**Pull request merged**